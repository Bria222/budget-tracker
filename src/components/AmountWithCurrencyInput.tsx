import * as React from "react";
import Grid from "@material-ui/core/Grid";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import { CurrencyInput } from "./CurrencyInput";
import { AmountInput } from "./AmountInput";
import { CurrencyRates } from "../api";
import applyRate from "../domain/utils/applyRate";
import { getCurrencyWithSymbol } from "../domain/utils/getCurrencyWithSymbol";
import { useLoc } from "../hooks/useLoc";

interface AmountCurrencyInputProps  {
    selectedCurrency: string;
    rates: CurrencyRates;
    amountInBaseCurrency?: number;
    onChange: (amount: number, currency: string, amountBase: number) => void;
    amountInput?: number;
    label: string;
    disabled?: boolean;
    onError?: (error?: string) => void;
}

export const AmountWithCurrencyInput: React.FC<AmountCurrencyInputProps> = (props) => {

    const loc  = useLoc();
    const [error, setError] = React.useState<string|undefined>(); 
    
    const {onChange, onError} = props;
    const {rates, base} = props.rates;

    function handleChange(amount: number, currency: string) {
        if (base && props.selectedCurrency && amount) {
            if (base === props.selectedCurrency) {
                onChange(amount, currency, amount);
            } else {
                try {
                    onChange(amount, currency, calculateAmountInBaseCurrency(amount, currency));
                    setError(undefined);
                } catch (error) {
                    setError(error);
                }
            }
        } else {
            onChange(amount, currency, amount);
        }
    }

    function calculateAmountInBaseCurrency(inputAmount: number, inputCurrency: string) {
        const rate = rates[inputCurrency];
        if (rate) {
            return applyRate(inputAmount, rate);
        } else {
            throw new Error(loc('Error fetching currencies'));
        }
    }

    React.useEffect(() => (onError && onError(error)), [error, onError]);

    const baseAmountString = () => {
        if (base !== props.selectedCurrency && props.amountInBaseCurrency) {
            return getCurrencyWithSymbol(props.amountInBaseCurrency, base);
        }
        return undefined;
    }

    function handleCurrencyChange (currency: string) {
        if (props.amountInput !== undefined) {
            handleChange(props.amountInput, currency);
        }
    }

    function handleAmountChange (amount: number) {
        if (props.selectedCurrency) {
            handleChange(amount, props.selectedCurrency);
        }
    }
    
    return (
        <Grid container direction='row' alignItems='baseline'>
            <Grid item>
                <AmountInput
                    amountInput={props.amountInput}
                    label={props.label}
                    onAmountChange={handleAmountChange}
                    helperText={baseAmountString()} 
                    disabled={props.disabled}/>
            </Grid>
            <Grid item >
                <CurrencyInput 
                    selectedCurrency={props.selectedCurrency}
                    onCurrencyChange={handleCurrencyChange} 
                    disabled={props.disabled}/>
            </Grid>
            { error !== undefined && 
            <Card>
                <CardContent>
                    {loc('Error converting to base"')}.
                    {loc('Connect to get last rates')}.
                    {loc('Still add amount in base')}.
                </CardContent>
            </Card> }
        </Grid>);
}

export default AmountWithCurrencyInput;
