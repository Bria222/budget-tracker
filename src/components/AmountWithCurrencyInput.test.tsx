import React from 'react';
import ReactDOM from 'react-dom';
import AmountWithCurrencyInput from './AmountWithCurrencyInput';
import renderer from 'react-test-renderer';

test('Amount Input changes', () => {
    const handleChange = (amount: number, currency: string, amountBase?: number) => {

    }
    const component = renderer.create(
        <AmountWithCurrencyInput 
            onChange={handleChange} 
            baseCurrency='EUR' 
            selectedCurrency='USD' />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    //tree.props.onMouseEnter();
    // re-rendering
    //tree = component.toJSON();
    //expect(tree).toMatchSnapshot();

    // manually trigger the callback
    //tree.props.onMouseLeave();
    // re-rendering
    //tree = component.toJSON();
    //expect(tree).toMatchSnapshot();
});