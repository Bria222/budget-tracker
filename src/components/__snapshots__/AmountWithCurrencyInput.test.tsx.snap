// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount Input changes 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense"
      style={
        Object {
          "margin": 8,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
        data-shrink={false}
        htmlFor="input-field-Amount"
      >
        Amount
        <span
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          aria-required={true}
          autoFocus={true}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          disabled={false}
          id="input-field-Amount"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          step=".01"
          type="number"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense"
      style={
        Object {
          "margin": 8,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled Mui-required Mui-required"
        data-shrink={true}
        htmlFor="input-field-Currency"
      >
        Currency
        <span
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root"
        >
          <select
            aria-invalid={false}
            className="MuiSelect-select MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            disabled={false}
            id="input-field-Currency"
            multiple={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            value="USD"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
  >
    <div
      className="MuiCardContent-root"
    >
      Error calculating amount in base currency. You need to be connected to Internet to get last currency rates. You can still add the amount in budget base currency.
    </div>
  </div>
</div>
`;
